### PR TITLE
Configure export command

### PR DESCRIPTION
**Pull request by msftrubengu**
This PR introduces the `configure export` command as an exprimental feature. This is mostly a proof of concept and should not be considered a full feature. 

#### Scenario A - Create a configuration to install a winget package.
Use `--pacakgeId` with the package identifier of an application in the winget source to produce a configuration file that uses the `Microsoft.WinGet.DSC/WinGetPackage` resource to install the package via winget. Right now, we don't validate if the package id exists and will just copy what the user sets into the settings of the resource.

#### Scenario B - 'Export' the configuration from the specified resource.
Use both `--module` and `--resource` to get the configuration of a resource and add it to the configuration file. Internally, configuration will install the module (if not installed already) and call Get on the resource. We will then try to serialize its property into the configuration file. Current limitation is that if a resource has a required setting (like WinGetPackage required Id) it won't work. If the resource is not found in the gallery a retry will be performed allowing prereleased modules in the case it exists.

#### Scenario C - Mix of A and B
If `--packageId`, `--module` and `--resource` are used, configure export will produce two resources. The first one is the `WinGetPackage` for the specified package. The second one is the same as in B, with the difference that it includes a dependency of the previously created `WinGetPackage` resource. 

#### Scenario D - Configuration file already exists.
If the file passed to the `--output` parameters already exists and is a valid configuration file, the resources will be appended. There is currently not validation into the correctness of this, so it can result in a configuration with resources with the same id.

For example `winget configure export --packageId Microsoft.AppInstaller  --module Microsoft.WinGet.DSC --resource WinGetUserSettings -o test_export.yml` would produce the following file

```
# Created using winget configure export 1.8.0-preview
# yaml-language-server: $schema=https://aka.ms/configuration-dsc-schema/0.2
properties:
  configurationVersion: 0.2
  resources:
  - resource: Microsoft.WinGet.DSC/WinGetPackage
    id: Microsoft.AppInstaller
    directives:
      description: Install Microsoft.AppInstaller
      allowPrerelease: true
    settings:
      id: Microsoft.AppInstaller
      source: winget
  - resource: Microsoft.WinGet.DSC/WinGetUserSettings
    dependsOn:
    - Microsoft.AppInstaller
    directives:
      description: Configure Microsoft.AppInstaller
    settings:
      Settings:
        experimentalFeatures:
          configureSelfElevate: true
        installBehavior:
          preferences:
            locale:
            - en-US
            - fr-FR
        $schema: https://aka.ms/winget-settings.schema.json
      Action: Full
```

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-cli/pull/4434)

### CODE REVIEW

File: src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp [hunk 1]
```diff
@@ -837,17 +846,264 @@ namespace AppInstaller::CLI::Workflow
...
+        }
     }
 
     void CreateConfigurationProcessor(Context& context)
     {
         auto progressScope = context.Reporter.BeginAsyncProgress(true);
         progressScope->Callback().SetProgressMessage(Resource::String::ConfigurationInitializing());
 
-        ConfigurationProcessor processor{ CreateConfigurationSetProcessorFactory(context)};
+        ConfigurationProcessor processor{ anon::CreateConfigurationSetProcessorFactory(context)};
```
Comment (Trenly) [inline src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp:1103]: Is there a reason for explicitly naming the anonymous namespace? I haven't seen this done before

Comment (JohnMcPMS) [reply src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp:1103]: I've learned that anonymous namespace does not output type information into the PDB in a way that debuggers (at least VS and WinDBG) can find currently, making it much harder to debug. With a named namespace, the derived type information is found and so the debugger can show the member fields of a derived type (vtable -> type mapping). One can manually do all of this with enough effort, and maybe eventually the debuggers will support it directly, but at this time I see simply naming the namespace as the easiest answer.

File: src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp [hunk 2]
```diff
@@ -837,17 +846,264 @@ namespace AppInstaller::CLI::Workflow
...
+                getResult = getAction.get();
+            }
+
+            progressScope.reset();
+            return getResult;
+        }
+
+        std::optional<ConfigurationUnit> CreateConfigurationUnit(Execution::Context& context, const std::optional<ConfigurationUnit> dependantUnit)
+        {
+            if (context.Args.Contains(Execution::Args::Type::ConfigurationExportModule) &&
```
[the remark this answers is not in the document]
Comment (msftrubengu) [reply src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp]: That was fun. I kept the name as Contains, tho.

File: src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp [hunk 3]
```diff
@@ -837,17 +846,264 @@ namespace AppInstaller::CLI::Workflow
...
+
+                ValueSet directives;
+                directives.Insert(s_Directive_Module, PropertyValue::CreateString(s_Module_WinGetClient));
+                directives.Insert(s_Directive_Description, PropertyValue::CreateString(L"Install " + packageIdWide));
+                directives.Insert(s_Directive_AllowPrerelease, PropertyValue::CreateBoolean(true));
+                unit.Metadata(directives);
+
+                ValueSet settings;
+                settings.Insert(s_Setting_Id, PropertyValue::CreateString(packageIdWide));
+                settings.Insert(s_Setting_Source, PropertyValue::CreateString(s_WinGetSource));
```
[the remark this answers is not in the document]
Comment (msftrubengu) [reply src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp:985]: I don't disagree, but that's a feature request or at least another PR.

File: src/Microsoft.Management.Configuration.Processor/Extensions/HashtableExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        /// </summary>
+        /// <param name="hashtable">hashtable.</param>
+        /// <returns>Value set.</returns>
+        public static ValueSet ToValueSet(this Hashtable hashtable)
+        {
+            var valueSet = new ValueSet();
+
+            foreach (DictionaryEntry entry in hashtable)
+            {
+                if (entry.Value == null)
```
[the remark this answers is not in the document]
Comment (msftrubengu) [reply src/Microsoft.Management.Configuration.Processor/Extensions/HashtableExtensions.cs]: I'm adding properties with null values now. However, I will ignore properties with null values doing serialization

File: src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Trenly) [reply src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp]: Thats one hell of a party trick; I love it!

[the remark this answers is not in the document]
Comment (denelon) [reply src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp]: 🤣

File: src/AppInstallerCLICore/ExecutionArgs.h [hunk 5]
```diff
@@ -162,7 +162,11 @@ namespace AppInstaller::CLI::Execution
             Max
         };
 
-        bool Contains(Type arg) const { return (m_parsedArgs.count(arg) != 0); }
+        template<typename... T, std::enable_if_t<(... && std::is_same_v<T, Args::Type>), bool> = true>
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/ExecutionArgs.h:165]: Pretty sure you can just do:
```suggestion
        template<Args::Type... T>
```

File: src/AppInstallerCLICore/ExecutionArgs.h [hunk 6]
```diff
@@ -162,7 +162,11 @@ namespace AppInstaller::CLI::Execution
             Max
         };
 
-        bool Contains(Type arg) const { return (m_parsedArgs.count(arg) != 0); }
+        template<typename... T, std::enable_if_t<(... && std::is_same_v<T, Args::Type>), bool> = true>
+        bool Contains(T... arg) const
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/ExecutionArgs.h:166]: You could also use a `std::initializer_list` with the low cost of putting `{}` around your parameter list.  But it definitely wouldn't be as fun as figuring out this.

File: src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp [hunk 7]
```diff
@@ -1112,13 +1111,14 @@ namespace AppInstaller::CLI::Workflow
                 directives.Insert(s_Directive_Module, PropertyValue::CreateString(moduleNameWide));
 
                 winrt::hstring description;
+                auto localized = std::wstring{ Resource::String::ConfigureExportUnitDescription };
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Workflows/ConfigurationFlow.cpp:1114]: If you are going to support localization here, you need to use the format specifier and the localization formatting functionality.  You can't just append strings together.

Also, you didn't do it for the `Install` above.

**Review by JohnMcPMS — APPROVED** [review]

**Pull request merged**